aws-lc-rs/src/kem.rs

```diff
@@ -0,0 +1,468 @@
+    priv_key: Box<[u8]>,
+}
+
+impl KemPrivateKey {
+    /// Generate a new KEM private key for the given algorithm.
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    pub fn generate(alg: &'static KemAlgorithm) -> Result<Self, Unspecified> {
+        unsafe {
+            let kyber_key = kem_key_generate(alg.id.nid())?;
+            let mut secret_key_size = alg.secret_key_size();
+            let mut priv_key_bytes = vec![0u8; secret_key_size];
+            if 1 != EVP_PKEY_get_raw_private_key(
+                kyber_key.as_const_ptr(),
+                priv_key_bytes.as_mut_ptr(),
+                &mut secret_key_size,
+            ) {
+                return Err(Unspecified);
+            }
+            Ok(KemPrivateKey {
+                algorithm: alg,
+                pkey: kyber_key,
+                priv_key: priv_key_bytes.into(),
+            })
+        }
+    }
+
+    /// Return the algorithm associated with the given KEM private key.
+    #[must_use]
+    pub fn algorithm(&self) -> &'static KemAlgorithm {
+        self.algorithm
+    }
+
+    /// Computes the KEM public key from the KEM private key
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    pub fn compute_public_key(&self) -> Result<KemPublicKey, Unspecified> {
+        let mut public_key_size = self.algorithm.public_key_size();
+        let mut pubkey_bytes = vec![0u8; public_key_size];
+        unsafe {
+            if 1 != EVP_PKEY_get_raw_public_key(
+                self.pkey.as_const_ptr(),
+                pubkey_bytes.as_mut_ptr(),
+                &mut public_key_size,
+            ) {
+                return Err(Unspecified);
+            }
+
+            let pubkey = LcPtr::new(EVP_PKEY_kem_new_raw_public_key(
+                self.algorithm.id.nid(),
+                pubkey_bytes.as_ptr(),
+                public_key_size,
+            ))?;
+
+            Ok(KemPublicKey {
+                algorithm: self.algorithm,
+                pkey: pubkey,
+                pub_key: pubkey_bytes.into(),
+            })
+        }
+    }
+
+    /// Performs the decapsulate operation using the current KEM private key on the given ciphertext.
+    ///
+    /// `ciphertext` is the ciphertext generated by the encapsulate operation using the KEM public key
+    /// associated with the current KEM private key.
+    ///
+    /// After the decapsulation is finished, `decapsulate` calls `kdf` with the raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    ///
+    /// # Errors
+    /// `error_value` when operation fails due to internal error.
+    ///
+    #[allow(clippy::missing_panics_doc)]
+    pub fn decapsulate<F, R, E>(
+        &self,
+        ciphertext: &mut [u8],
+        error_value: E,
+        kdf: F,
+    ) -> Result<R, E>
+    where
+        F: FnOnce(&[u8]) -> Result<R, E>,
+    {
+        unsafe {
```

Comment:
Try to reduce the scopes of your unsafe blocks to the smallest portions as possible. Most of the code here isn't unsafe.